subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultFileCollectionResolveContext.java

```diff
@@ -102,50 +103,57 @@ public final ResolvableFileCollectionResolveContext newContext() {
-                    queue.add(0, deferredResult);
-                }
-            } else if (element instanceof Path) {
-                queue.add(0, ((Path) element).toFile());
-            } else if (element instanceof Iterable) {
-                Iterable<?> iterable = (Iterable) element;
-                GUtil.addToCollection(queue.subList(0, 0), iterable);
-            } else if (element instanceof Object[]) {
-                Object[] array = (Object[]) element;
-                GUtil.addToCollection(queue.subList(0, 0), Arrays.asList(array));
-            } else {
+            Object element = queue.remove();
+            resolveElement(converter, result, element);
+        }
+        return result;
+    }
+
+    // TODO - need to sync with BuildDependenciesOnlyFileCollectionResolveContext
+    private <T> void resolveElement(Converter<? extends T> converter, List<T> result, Object element) {
+        if (element instanceof DefaultFileCollectionResolveContext) {
+            converter.convertInto(element, result, fileResolver);
+        } else if (element instanceof FileCollectionContainer) {
+            resolveNested(converter, result, (FileCollectionContainer) element);
+        } else if (element instanceof FileCollection || element instanceof MinimalFileCollection) {
+            converter.convertInto(element, result, fileResolver);
+        } else if (element instanceof Task) {
+            resolveElement(converter, result, ((Task) element).getOutputs().getFiles());
+        } else if (element instanceof TaskOutputs) {
+            resolveElement(converter, result, ((TaskOutputs) element).getFiles());
+        } else if (DeferredUtil.isDeferred(element)) {
+            resolveElement(converter, result, DeferredUtil.unpack(element));
+        } else if (element instanceof Path) {
+            resolveElement(converter, result, ((Path) element).toFile());
+        } else {
+            if (element instanceof Object[]) {
+                element = Arrays.asList((Object[]) element);
+            }
+
+            if (element instanceof Iterable) {
+                resolveElements(converter, result, (Iterable<?>) element);
+            } else if (element != null) {
                 converter.convertInto(element, result, fileResolver);
             }
         }
-        return result;
     }
 
-    protected <T> void resolveNested(FileCollectionContainer fileCollection, List<T> result, Converter<? extends T> converter) {
-        addTo = queue.subList(0, 0);
+    private <T> void resolveElements(Converter<? extends T> converter, List<T> result, Iterable<?> element) {
+        for (Object elem : element) {
+            resolveElement(converter, result, elem);
+        }
+    }
+
+    private <T> void resolveNested(Converter<? extends T> converter, List<T> result, FileCollectionContainer element) {
+        Queue<Object> copy = queue;
         try {
-            fileCollection.visitContents(this);
+            queue = new ArrayDeque<>();
```

Comment:
Could we get rid of this state?

Reply:
Would be neat, it's the only reason I've kept the queue: for the `visitContents` callbacks.
Do you have any specific recommendation?

Reply:
I mainly meant the fact that we reassign the queue field instead of passing it as an argument for instance or creating a new context object.

Reply:
Yes, wanted to do it, but ended up changing too many things. Can you think of a way to do that nicely?

Reply:
Create a new context instead of passing this.

Reply:
I checked it previously, but since it requires a FileCollectionResolveContext (and didn't understand how `push` is related to the flow), decided to use the old swapping technique - if you can explain it to me, I'll be happy to attempt it in a different PR (or this one, if it doesn't complicate it).
Feel free to try it yourself, I don't mind if you push to this PR.

Reply:
Done